Mewbot discord dice roller example plugin

ajCameron opened the pull request:
- aims to resolve https://github.com/mewbotorg/mewbot/issues/182
 - minor correction to the windows dev install docs
 - added example of direct import - access through registry needs more work than should be accommodated in this pull request
 - added mewbot-discord_dice_roller plugin with example
 - upgraded examples to allow running of examples stored in builtin plugins via path manipulation
 - added an "examples" sh script - through which you can run the examples
 - added notes to API_DISTRIBUTIONS constant in registry
 - fixed weird issue in bot which was preventing run - on windows - with some python versions

Still todo
 - sort out packaging for modules
 - write an interface for registry to allow direct listing and loading of modules from there
however, including these would grow this pull request beyond reasonable scope

ajCameron committed:
Adding basic README.md for basic plugin development

 - describing the types of plugin you might want to write

ajCameron committed:
Re-adding the discord-dice-roller plugin

 - aims to resolve https://github.com/mewbotorg/mewbot/issues/182
 - minor correction to the windows dev install docs
 - added example of direct import - access through registry needs more work than should be accommodated in this pull request
 - added mewbot-discord_dice_roller plugin with example
 - upgraded examples to allow running of examples stored in builtin plugins via path manipulation
 - added an "examples" sh script - through which you can run the examples
 - added notes to API_DISTRIBUTIONS constant in registry
 - fixed weird issue in bot which was preventing run - on windows - with some python versions

exachixkitsune commented on dev-docs/plugin-dev-notes/README.md:
````diff
@@ -0,0 +1,106 @@
+
+#### Example 1
+
+"I want to allow mewbot to talk to mastadon"
+
+Probably write a **framework** plugin.
+
+In fact, please do!
+That would really be very helpful.
+
+Your generic IOConfig will take the form of a namespace plugin.
+It will be used by importing it using something like
+
+```python
+from mewbot.io.mastadon import MastadonIOConfig
+```
````
```suggestion
"I want to allow mewbot to talk to mastodon"

Probably write a **framework** plugin.

In fact, please do!
That would really be very helpful.

Your generic IOConfig will take the form of a namespace plugin.
It will be used by importing it using something like

```python
from mewbot.io.mastodon import MastodonIOConfig
```
```

ajCameron replied:
I've gone through and re-written this section to be more definite "probably write" has become "write" - in line with the suggestion below.

exachixkitsune commented on dev-docs/plugin-dev-notes/README.md:
```diff
@@ -0,0 +1,106 @@
+
+"I've written a discord dice roller! I'd like to distribute it."
+
+Probably write a **bot** plugin.
```
```suggestion
Write a **bot** plugin.
```

It is my opinion these should be definitives rather than have wiggle room; but that's potentially on me.
Would there be a reason we make a discord dice roller as part of the framework vs a plugin?

ajCameron replied:
Agreed - implemented throughout the docs.

exachixkitsune commented on plugins/mewbot-discord_dice_roller/README.md:
```diff
@@ -0,0 +1,42 @@
+
+Which of these you use is up to you!
+
+Included in this module is two examples - one for each method.
```
I would also include what needs to go into the `token: "[token goes here]"` part of the `direct_import_dice_roller_bot.yaml` file.

ajCameron replied:
Included and referenced in the places which seem to make sense.

exachixkitsune commented on plugins/mewbot-discord_dice_roller/setup.py:
```diff
@@ -0,0 +1,75 @@
+It _should_ also work for local installation, but this functionality is not tested.
+"""
+
+raise NotImplementedError("Still under development.")
```
Do we want this implemented for the purpose of this demo? or is it dependant on other things?

javajawa replied:
It's...a little complicated. We're already doing some pretty nonstandard things with `setup.py` in other places with the multi-module build system. Probably could make it work right now, but I'm happy to ignore it.

I'm actually going to suggest (when we preoperly get there) using `pyproject.toml` rather than `setup.py` for the plugins -- that uses more sensible defaults and means there isn't a random python file for installation.

ajCameron replied:
I'm gonna have a think and a poke some, and see what I come up with.

ajCameron replied:
I have something which works, but feel it's outside the scope of this pull request.

exachixkitsune commented on plugins/mewbot-discord_dice_roller/src/mewbot_discord_dice_roller/behaviors.py:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+
+"""
+This Behavior brings all the componets for the Discord Dice roller together in one place.
```
```suggestion
This Behavior brings all the components for the Discord Dice roller together in one place.
```

ajCameron replied:
Corrected

exachixkitsune commented on plugins/mewbot-discord_dice_roller/src/mewbot_discord_dice_roller/triggers.py:
```diff
@@ -0,0 +1,45 @@
+        :return:
+        """
+
+        if not isinstance(event, DiscordMessageCreationEvent):
```
Idle thought for future code; is there a way to make sure this happens automatically/is not forgettable by a user?
Is there a reason not to do that?

javajawa replied:
The [Liskov substitution principle](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides) boils down to the argument to the function being `InputEvent` (or `OutputEvent` for `Output`s), so the check is going to be needed for typing purposes. It's also just a good defensive measure against poorly implemented bot runners.

It _is_ a little cumbersome, and we could abstract it to a decorator...
A fully generic version might look like

```python
from typing import Callable, ParamSpec, TypeVar

import mewbot.api.v1
from mewbot.core import InputEvent, OutputEvent
from mewbot.io.common import EventWithReplyMixIn


Args = ParamSpec("Args")
Bound = TypeVar("Bound", bound=InputEvent)
Return = TypeVar("Return")


def check_types(*types: type[Bound], result: Return = ...) -> Callable[[Callable[Args, Return]], Callable[Args, Return]]:
    def check_wrapper(wrapped: Callable[Args, Return]) -> Callable[Args, Return]:
        def inner(*args: Args.args, **kwargs: Args.kwargs) -> Return:
            print(args)
            if not isinstance(args[1], types):
                if result is ...:
                    raise TypeError(f"First argument to {wrapped} must be one of {types}")
                return result

            return wrapped(*args, **kwargs)

        return inner
    return check_wrapper


class A(mewbot.api.v1.Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {EventWithReplyMixIn}

    @check_types(EventWithReplyMixIn, result=False)
    def matches(self, event: InputEvent) -> bool:
        return True
```

I suspect one which is configured specifically for the different functions would, however, be more useful (and not make mypy scream)

... ok, I took some time to play and it is 100% possible:

```python
from __future__ import annotations

from typing import Callable, TypeVar

from mewbot.api.v1 import Trigger, Condition, InputEvent
from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent

Component = TypeVar("Component", bound=Trigger | Condition)
Event = TypeVar("Event", bound=InputEvent)


def check_types(wrapped: Callable[[Component, Event], bool]) -> Callable[[Component, InputEvent], bool]:
    def inner(self: Component, event: InputEvent) -> bool:
        if not isinstance(event, tuple(self.consumes_inputs())):
            return False

        return wrapped(self, event)  # type: ignore

    return inner


class A(Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {DiscordInputEvent}

    @check_types
    def matches(self, event: DiscordInputEvent) -> bool:
        print(event)
        return True


print("invoking A()")
print(A().matches(InputEvent()))
print(A().matches(DiscordInputEvent()))
print(A().matches(DiscordUserJoinInputEvent("nya")))
```

That passes typing checks; there is one ignore there because I can't _prove_ that `Bound` is implied by the `isintance` call above.

javajawa replied:
...if we want to get really, really clever, we could actually shift the concern of `consume_inputs` to the type signature of `matches`...

I think this is where it veers into bad idea territory. And I can't easily make it work, but I have made a slightly better version of the above (which doesn't need any mypy suppressions, but *does* confuse PyCharm a bit):
```python
from __future__ import annotations

import functools

from types import UnionType
from typing import Callable, TypeVar, get_type_hints

from mewbot.api.v1 import Trigger, Condition, InputEvent

Component = TypeVar("Component", bound=Trigger | Condition)
Event = TypeVar("Event", bound=InputEvent)


def check_types(
    wrapped: Callable[[Component, Event], bool]
) -> Callable[[Component, InputEvent], bool]:
    types = get_type_hints(wrapped)
    if "event" not in types:
        raise TypeError("Received function without 'event' parameter")

    event_types: type[Event] = types["event"]
    if not isinstance(event_types, (type, UnionType)):
        raise TypeError("Received function with invalid type for 'event'")

    @functools.wraps(wrapped)
    def inner(self: Component, event: InputEvent) -> bool:
        # noinspection PyTypeHints
        if not isinstance(event, event_types):
            return False

        return wrapped(self, event)

    return inner


# Demo Code

from mewbot.io.discord import DiscordInputEvent, DiscordUserJoinInputEvent
from mewbot.io.http import IncomingWebhookEvent


class A(Trigger):
    @staticmethod
    def consumes_inputs() -> set[type[InputEvent]]:
        return {DiscordInputEvent, IncomingWebhookEvent}

    @check_types
    def matches(self, event: DiscordInputEvent | IncomingWebhookEvent) -> bool:
        """Hello I am documentation"""
        return True


print(A.matches)  # <function A.matches at 0x7f6e703f4a40>
print(A.matches.__doc__)  # <function A.matches at 0x7f6e703f4a40>
print(A().matches(InputEvent()))  # False
print(A().matches(DiscordInputEvent()))  # True
print(A().matches(DiscordUserJoinInputEvent("nya")))  # True
```

And I think we could store the `UnionType` somewhere and use it for `consume_inputs` if we really wanted to. But I also feel that might be Too Much Magic:tm:

ajCameron replied:
I really like this. It's very elegant and could be more magical.

This feels like it should be it's own pull request. I'll do one tomorrow to get it into the code base and get it integrated.

exachixkitsune reviewed (dismissed):
Looking good, comments are mostly ignorable, spelling, or idle future thoughts.

ajCameron committed:
Documentation and spelling corrections and upgrades

 - written a section in discord-dev-notes.md concerning how to actually run the discord example bots
 - linux is still a placeholder, but it should be pretty easy to work out from the windows version
 - re-written docs to be affirmative ("maybe" and "perhaps" replace with clear instructions)
 - included references to the instructions as to how to run the discord examples in the plugin __init__.py and several places in the docs
 - correcting spelling error in behaviors.py

exachixkitsune reviewed (approved):
Looks good! Extent conversations - probably need you to decide if they are solved, or deferred to other PRs.
